File: 002.go
package main

import "fmt"

func main() {
  total, term1, fib := 0
  term2 := 1

  for {
    fib = term1 + term2
    if fib % 2 == 0 {
      total += fib
    }
    term2 = term1
    term1 = fib
   
    if fib > 4000000 {
      break
    }
  }
  fmt.Println(total)
}

File: 002.hs
fib :: Int -> Int
fib 1 = 1
fib 2 = 2
fib n = fib(n - 1) + fib(n - 2)

isEven n = if mod n 2 == 0 then True else False

-- Slow because we re-compute fib for each number.
total = sum (filter isEven (takeWhile (< 4000000) (map fib [1..])))

main = do
  putStrLn(show total)
